File: apps/api/app/schemas/billing.py
from datetime import datetime

from pydantic import BaseModel


class BillingCaseCreate(BaseModel):
    patient_id: str
    appointment_id: str | None = None
    payer: str | None = None
    cpt_codes: list[str] = []
    diagnosis_codes: list[str] = []
    notes: str | None = None


class BillingCaseUpdate(BaseModel):
    status: str | None = None
    payer: str | None = None
    eligibility_status: str | None = None
    claim_control_number: str | None = None
    remittance_status: str | None = None
    allowed_amount: float | None = None
    paid_amount: float | None = None
    cpt_codes: list[str] | None = None
    diagnosis_codes: list[str] | None = None
    notes: str | None = None


class BillingCaseOut(BaseModel):
    id: str
    patient_id: str
    appointment_id: str | None
    status: str
    payer: str | None
    eligibility_status: str
    claim_control_number: str | None
    submission_ready_at: datetime | None
    submitted_at: datetime | None
    denied_at: datetime | None
    denial_reason: str | None
    denial_worked_at: datetime | None
    remittance_status: str
    allowed_amount: float | None
    paid_amount: float | None
    paid_at: datetime | None
    cpt_codes: list[str]
    diagnosis_codes: list[str]
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BillingCaseListOut(BaseModel):
    data: list[BillingCaseOut]
    total: int


class ChargeReviewItemOut(BaseModel):
    encounter_id: str
    patient_id: str
    patient_name: str
    appointment_id: str | None
    encounter_type: str
    signed_at: datetime | None
    summary: str | None
    recommended_cpt_codes: list[str]
    recommended_diagnosis_codes: list[str]


class ChargeReviewListOut(BaseModel):
    data: list[ChargeReviewItemOut]
    total: int


class EligibilityCheckOut(BaseModel):
    patient_id: str
    payer: str | None
    status: str
    reference_id: str
    message: str


class BillingTimelineEventOut(BaseModel):
    id: str
    source: str = "audit"
    event_type: str
    entity_type: str
    entity_id: str
    actor_id: str | None
    payload: dict
    created_at: datetime
    status: str | None = None

    model_config = {"from_attributes": True}


class BillingTimelineOut(BaseModel):
    data: list[BillingTimelineEventOut]
    total: int


class BillingClaimReadinessOut(BaseModel):
    case_id: str
    ready: bool
    blockers: list[str]
    warnings: list[str]
    recommended_next_step: str


class BillingWorkQueueOut(BaseModel):
    draft_count: int
    ready_count: int
    submitted_count: int
    denied_count: int
    paid_count: int
    missing_coding_count: int
    eligibility_needed_count: int
    denial_rework_count: int
    remittance_pending_count: int
    total: int


class BillingPaymentIn(BaseModel):
    allowed_amount: float | None = None
    paid_amount: float | None = None
    remittance_status: str | None = None
    notes: str | None = None


class BillingReworkIn(BaseModel):
    notes: str | None = None

File: packages/shared/src/types/billing.ts
import type { UUID } from './common';

export type BillingCaseStatus = 'draft' | 'ready' | 'submitted' | 'denied' | 'paid';

export interface BillingCase {
  id: UUID;
  patient_id: UUID;
  appointment_id: UUID | null;
  status: BillingCaseStatus;
  payer: string | null;
  eligibility_status: string;
  claim_control_number: string | null;
  submission_ready_at: string | null;
  submitted_at: string | null;
  denied_at: string | null;
  denial_reason: string | null;
  denial_worked_at: string | null;
  remittance_status: string;
  allowed_amount: number | null;
  paid_amount: number | null;
  paid_at: string | null;
  cpt_codes: string[];
  diagnosis_codes: string[];
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface BillingCaseListResponse {
  data: BillingCase[];
  total: number;
}

export interface ChargeReviewItem {
  encounter_id: UUID;
  patient_id: UUID;
  patient_name: string;
  appointment_id: UUID | null;
  encounter_type: string;
  signed_at: string | null;
  summary: string | null;
  recommended_cpt_codes: string[];
  recommended_diagnosis_codes: string[];
}

export interface ChargeReviewListResponse {
  data: ChargeReviewItem[];
  total: number;
}

export interface EligibilityCheck {
  patient_id: UUID;
  payer: string | null;
  status: string;
  reference_id: string;
  message: string;
}

export interface BillingTimelineEvent {
  id: UUID;
  source?: 'audit' | 'integration';
  event_type: string;
  entity_type: string;
  entity_id: UUID;
  actor_id: UUID | null;
  payload: Record<string, unknown>;
  created_at: string;
  status?: string | null;
}

export interface BillingTimelineResponse {
  data: BillingTimelineEvent[];
  total: number;
}

export interface BillingClaimReadiness {
  case_id: UUID;
  ready: boolean;
  blockers: string[];
  warnings: string[];
  recommended_next_step: string;
}

export interface BillingWorkQueue {
  draft_count: number;
  ready_count: number;
  submitted_count: number;
  denied_count: number;
  paid_count: number;
  missing_coding_count: number;
  eligibility_needed_count: number;
  denial_rework_count: number;
  remittance_pending_count: number;
  total: number;
}
